File: clojure/src/optikpi/datapipeline/models/account_event.clj
(ns optikpi.datapipeline.models.account-event)

(defn validate [m]
  (let [errors
        (cond-> []
          (empty? (:account_id m))   (conj "account_id is required")
          (empty? (:workspace_id m)) (conj "workspace_id is required")
          (empty? (:user_id m))      (conj "user_id is required")
          (empty? (:event_name m))   (conj "event_name is required")
          (empty? (:event_id m))     (conj "event_id is required")
          (empty? (:event_time m))   (conj "event_time is required")
          (and (seq (:event_category m))
               (not= (:event_category m) "Account"))
          (conj "event_category must be \"Account\" for account events")
          (and (some? (:device m))
               (not (#{"desktop" "mobile" "tablet" "app"} (:device m))))
          (conj "device must be one of: desktop, mobile, tablet, app")
          (and (some? (:event_time m))
               (not (try (java.time.Instant/parse (:event_time m)) true (catch Exception _ false))))
          (conj "event_time must be in ISO 8601 format (YYYY-MM-DDTHH:mm:ssZ)"))]
    {:valid? (empty? errors) :errors errors}))

(defn build [data]
  (assoc data :event_category "Account"))

File: rust/src/models/account_event.rs
use serde::{Deserialize, Serialize};

use super::{is_valid_datetime, ok, ValidationResult};

const VALID_DEVICES: [&str; 4] = ["desktop", "mobile", "tablet", "app"];

/// Account event: registration, verification, and account changes.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccountEvent {
    pub account_id: String,
    pub workspace_id: String,
    pub user_id: String,
    pub event_category: String,
    pub event_name: String,
    pub event_id: String,
    pub event_time: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub device: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub affiliate_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub partner_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub campaign_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub utm_source: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub utm_campaign: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub utm_medium: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

impl AccountEvent {
    pub fn new(
        account_id: impl Into<String>,
        workspace_id: impl Into<String>,
        user_id: impl Into<String>,
        event_name: impl Into<String>,
        event_id: impl Into<String>,
        event_time: impl Into<String>,
    ) -> Self {
        Self {
            account_id: account_id.into(),
            workspace_id: workspace_id.into(),
            user_id: user_id.into(),
            event_category: "Account".to_string(),
            event_name: event_name.into(),
            event_id: event_id.into(),
            event_time: event_time.into(),
            device: None,
            status: None,
            affiliate_id: None,
            partner_id: None,
            campaign_code: None,
            utm_source: None,
            utm_campaign: None,
            utm_medium: None,
            reason: None,
        }
    }

    pub fn validate(&self) -> ValidationResult {
        let mut errors = Vec::new();

        if self.account_id.is_empty() {
            errors.push("account_id is required".to_string());
        }
        if self.workspace_id.is_empty() {
            errors.push("workspace_id is required".to_string());
        }
        if self.user_id.is_empty() {
            errors.push("user_id is required".to_string());
        }
        if self.event_name.is_empty() {
            errors.push("event_name is required".to_string());
        }
        if self.event_id.is_empty() {
            errors.push("event_id is required".to_string());
        }
        if self.event_time.is_empty() {
            errors.push("event_time is required".to_string());
        }
        if self.event_category != "Account" {
            errors.push(r#"event_category must be "Account" for account events"#.to_string());
        }
        if let Some(device) = &self.device {
            if !VALID_DEVICES.contains(&device.as_str()) {
                errors.push(format!("device must be one of: {}", VALID_DEVICES.join(", ")));
            }
        }
        if !self.event_time.is_empty() && !is_valid_datetime(&self.event_time) {
            errors.push("event_time must be in ISO 8601 format (YYYY-MM-DDTHH:mm:ssZ)".to_string());
        }

        ok(errors)
    }
}
